geps/gep-3779/index.md

```diff
@@ -42,20 +37,429 @@ Kubernetes core provides NetworkPolicies as one way to do it. Network Policies h
+    // +required
+    Type AuthorizationSourceType `json:"type"`
+
+    // spiffe specifies an identity that is matched by this rule.
+    //
+    // spiffe identities must be specified as SPIFFE-formatted URIs following the pattern:
+    //   spiffe://<trust_domain>/<workload-identifier>
+    //
+    // While the exact workload identifier structure is implementation-specific,
+    // implementations are encouraged to follow the convention of
+    // `spiffe://<trust_domain>/ns/<namespace>/sa/<serviceaccount>`
+    // when representing Kubernetes workload identities.
+    //
+    // spiffe identities for authorization can be derived in various ways by the underlying
+    // implementation. Common methods include:
+    // - From peer mTLS certificates: The identity is extracted from the client's
+    //   mTLS certificate presented during connection establishment.
+    // - From IP-to-identity mappings: The implementation might maintain a dynamic
+    //   mapping between source IP addresses (pod IPs) and their associated
+    //   identities (e.g., Service Account, SPIFFE IDs).
+    // - From JWTs or other request-level authentication tokens.
+    //
+    // Note for reviewers: While this GEP primarily focuses on identity-based
+    // authorization where identity is often established at the transport layer,
+    // some implementations might derive identity from authenticated tokens or sources
+    // within the request itself.
+    //
+    // +optional
+    SPIFFE *AuthorizationSourceSPIFFE `json:"spiffe,omitempty"`
+
+    // ServiceAccount specifies a Kubernetes Service Account that is
+    // matched by this rule. A request originating from a pod associated with
+    // this serviceaccount will match the rule.
+    //
+    // The ServiceAccount listed here is expected to exist within the same
+    // trust domain as the targeted workload, which in many environments means
+    // the same Kubernetes cluster. Cross-cluster or cross-trust-domain access
+    // should instead be expressed using the `SPIFFE` field.
+    // +optional
+    ServiceAccount AuthorizationSourceServiceAccount `json:"serviceAccount,omitempty"`
+}
+
+type AuthorizationSourceSPIFFE AbsoluteURI
+
+type AuthorizationSourceServiceAccount struct {
+  // Namespace is the namespace of the ServiceAccount
+  // If not specified, current namespace (the namespace of the policy) is used.
+  Namespace *Namespace `json:"namespace,omitempty"`
+
+  // Name is the name of the ServiceAccount. 
+  // Use "*" to indicate all serviceaccounts in the namespace.
+  // +required
+  Name string `json:"name"`
+}
+
+
+// AuthorizationAttribute defines L4 properties of a request destination.
+type AuthorizationNetworkAttributes struct {
+    // Ports specifies a list of destination ports to match on.
+    // Traffic is matched if it is going to any of these ports.
+    // If not specified, the rule applies to all ports.
+    // +optional
+    Ports []gatewayv1.PortNumber `json:"ports,omitempty"`
```

Comment:
I think there's some prior art either in an ancient thread in Gateway API or somewhere in https://github.com/kubernetes-sigs/network-policy-api/ about how port ranges (e.g. `5000-5200`) might be specified. An array of port numbers is _probably fine_ for an initial implementation, but I'd expect we might want a breaking change here before any potential advancement to standard unless we want to limit this to a singular `port` field to allow a backwards-compatible `ports` field to be added later.

Reply:
Are you familiar with any implementation that does port ranges? I have not encountered this use case too much

Reply:
In ClusterNetworkPolicy (recently renamed from AdminNetworkPolicy), `ports` is an array of [ClusterNetworkPolicyPort](https://network-policy-api.sigs.k8s.io/reference/spec/#policy.networking.k8s.io%2fv1alpha2.ClusterNetworkPolicyPort) objects, which can express a single `portNumber`, a `portRange`, or a `namedPort`

For use cases in AuthZ policy, some databases do weird things like have clients open connections on dynamic ports in a range https://stackoverflow.com/questions/49218476/connection-to-sql-server-uses-random-ports